### PR TITLE
Add Ktlint pre-commit hook

### PR DESCRIPTION
**Pull request by ShridharGoel**
## Purpose / Description
Add Git pre-commit hook for Ktlint.
If this is installed, then `./gradlew ktlintFormat` will run for Kotlin files and then `git add $file` will run on those files before commit.

## Fixes
Fixes part of #9019 

## Approach
Added pre-commit file and task to install Git pre-commit hook for Ktlint before build.

## How Has This Been Tested?
Verified with a test commit.

## Learning
https://medium.com/@alistair.cerio/android-ktlint-and-pre-commit-git-hook-5dd606e230a9
https://emmanuelkehinde.io/setting-up-git-pre-commit-pre-push-hook-for-ktlint-check

Reference for the pre-commit script: https://gist.github.com/echorebel/4246cd4ec42098fce66c169d28aa2812#gistcomment-3734796

## Checklist
- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

File: AnkiDroid/build.gradle [hunk 1]
```diff
@@ -224,6 +224,14 @@ tasks.withType(JavaCompile) {
     options.incremental = true
 }
 
+// Install Git pre-commit hook for Ktlint
+task installGitHook(type: Copy) {
+    from new File(rootProject.rootDir, 'pre-commit')
+    into { new File(rootProject.rootDir, '.git/hooks') }
+    fileMode 0777
```
Comment (mikehardy) [inline AnkiDroid/build.gradle:231]: Should not be world/group writable? 

```suggestion
    fileMode 0755
```

**Review by mikehardy — CHANGES_REQUESTED** [review]
One comment because I think file mode 777 is not recommended in any case for security reasons, but otherwise looks right - can you confirm you tested locally and flags badness :-) ?

**Comment by ShridharGoel** [comment]
> can you confirm you tested locally and flags badness :-) ?

Yes, I have tested it.

**Comment by david-allison-1** [comment]
```
0 file committed, 1 file failed to commit: Backups: use IncrementerNumberRangePreference Requested by review Running Git pre-commit hook for Ktlint > Task :loadKtlintReporters UP-TO-DATE > Task :runKtlintCheckOverKotlinScripts NO-SOURCE > Task :api:loadKtlintReporters FROM-CACHE > Task :lint-rules:loadKtlintReporters FROM-CACHE > Task :AnkiDroid:loadKtlintReporters FROM-CACHE > Task :ktlintKotlinScriptCheck SKIPPED > Task :ktlintCheck UP-TO-DATE > Task :AnkiDroid:runKtlintCheckOverAmazonDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAmazonDebugSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverAndroidTestDebugSourceSet NO-SOURCE > Task :api:ktlintAndroidTestDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAmazonReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAmazonReleaseSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverAndroidTestReleaseSourceSet NO-SOURCE > Task :api:ktlintAndroidTestReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAmazonSourceSet NO-SOURCE > Task :api:runKtlintCheckOverAndroidTestSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAmazonSourceSetCheck SKIPPED > Task :api:ktlintAndroidTestSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverDebugSourceSet NO-SOURCE > Task :AnkiDroid:runKtlintCheckOverAndroidTestAmazonDebugSourceSet NO-SOURCE > Task :api:ktlintDebugSourceSetCheck SKIPPED > Task :AnkiDroid:ktlintAndroidTestAmazonDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAndroidTestAmazonSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAndroidTestAmazonSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverKotlinScripts NO-SOURCE > Task :api:ktlintKotlinScriptCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAndroidTestDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAndroidTestDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAndroidTestPlayDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAndroidTestPlayDebugSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverMainSourceSet NO-SOURCE > Task :api:ktlintMainSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAndroidTestPlaySourceSet NO-SOURCE > Task :AnkiDroid:ktlintAndroidTestPlaySourceSetCheck SKIPPED > Task :api:runKtlintCheckOverReleaseSourceSet NO-SOURCE > Task :AnkiDroid:runKtlintCheckOverAndroidTestReleaseSourceSet NO-SOURCE > Task :api:ktlintReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:ktlintAndroidTestReleaseSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverTestDebugSourceSet NO-SOURCE > Task :api:ktlintTestDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverAndroidTestSourceSet NO-SOURCE > Task :AnkiDroid:ktlintAndroidTestSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverTestReleaseSourceSet NO-SOURCE > Task :api:ktlintTestReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintDebugSourceSetCheck SKIPPED > Task :api:runKtlintCheckOverTestSourceSet NO-SOURCE > Task :api:ktlintTestSourceSetCheck SKIPPED > Task :api:ktlintCheck UP-TO-DATE > Task :AnkiDroid:runKtlintCheckOverKotlinScripts NO-SOURCE > Task :AnkiDroid:ktlintKotlinScriptCheck SKIPPED > Task :lint-rules:runKtlintCheckOverKotlinScripts NO-SOURCE > Task :lint-rules:ktlintKotlinScriptCheck SKIPPED > Task :lint-rules:runKtlintCheckOverMainSourceSet NO-SOURCE > Task :lint-rules:ktlintMainSourceSetCheck SKIPPED > Task :lint-rules:runKtlintCheckOverTestSourceSet NO-SOURCE > Task :lint-rules:ktlintTestSourceSetCheck SKIPPED > Task :lint-rules:ktlintCheck UP-TO-DATE > Task :AnkiDroid:runKtlintCheckOverPlayDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintPlayDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverPlayReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintPlayReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverPlaySourceSet NO-SOURCE > Task :AnkiDroid:ktlintPlaySourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestAmazonDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestAmazonDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestAmazonReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestAmazonReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestAmazonSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestAmazonSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestPlayDebugSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestPlayDebugSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestPlayReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestPlayReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestPlaySourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestPlaySourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestReleaseSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestReleaseSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverTestSourceSet NO-SOURCE > Task :AnkiDroid:ktlintTestSourceSetCheck SKIPPED > Task :AnkiDroid:runKtlintCheckOverMainSourceSet > Task :AnkiDroid:ktlintMainSourceSetCheck FAILED C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\utils\UiUtil.kt:1:1 File must end with a newline (\n) C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\src\main\java\com\ichi2\utils\UiUtil.kt:26:40 Unnecessary space(s) FAILURE: Build failed with an exception. * What went wrong: Execution failed for task ':AnkiDroid:ktlintMainSourceSetCheck'. > A failure occurred while executing org.jlleitschuh.gradle.ktlint.worker.ConsoleReportWorkAction > KtLint found code style violations. Please see the following reports: - C:\GitHub\Anki-Android-Git\Anki-Android\AnkiDroid\build\reports\ktlint\ktlintMainSourceSetCheck\ktlintMainSourceSetCheck.txt * Try: Run with --stacktrace option to get the stack trace. Run with --info or --debug option to get more log output. Run with --scan to get full insights. * Get more help at https://help.gradle.org Deprecated Gradle features were used in this build, making it incompatible with Gradle 8.0. Use '--warning-mode all' to show the individual deprecation warnings. See https://docs.gradle.org/7.0.2/userguide/command_line_interface.html#sec:command_line_warnings BUILD FAILED in 8s 6 actionable tasks: 2 executed, 3 from cache, 1 up-to-date
```

* Could we improve the error? There's lots of noise
* @mikehardy - you're more experienced here. I feel there's an argument to `KtLintFormat` on pre-commit, rather than `KtLintCheck`

**Comment by david-allison-1** [comment]
This appears to have removed everything on failure

image

**Comment by ShridharGoel** [comment]
> * I feel there's an argument to `KtLintFormat` on pre-commit, rather than `KtLintCheck`

Yes, that can be done. That's what I wanted to ask in https://github.com/ankidroid/Anki-Android/issues/9019#issuecomment-851351009.

**Comment by ShridharGoel** [comment]
> * I feel there's an argument to `KtLintFormat` on pre-commit, rather than `KtLintCheck`

Yes, that can be done. That's what I wanted to ask in https://github.com/ankidroid/Anki-Android/issues/9019#issuecomment-851351009, about whether we should just check or should we format as well.

**Comment by ShridharGoel** [comment]
> This appears to have removed everything on failure
> 
> image

Did something go wrong because of the `stash` commands? Can you tell what exact steps did you follow?

**Comment by david-allison-1** [comment]
> > This appears to have removed everything on failure
> > image
> 
> Did something go wrong because of the `stash` commands? Can you tell what exact steps did you follow?

* git pr 9028 (personal script: fetch PR from GitHub & checkout)
* removed a couple lines from `UiUtils.kt` 
* commit in Android Studio

**Comment by ShridharGoel** [comment]
Updated to run `./gradlew ktlintFormat` on Kotlin files and then run `git add $file` on the files before commit.

**Comment by ShridharGoel** [comment]
@david-allison-1 Can you try again?

**Comment by david-allison-1** [comment]
at b82ec64db69ddf3bde8c211e4404ca6a4368e1bd

* git pr 9028 2
* Android Studio - Rebuild
* Make changes to UiUtil
* Rebuild

Same thing - all files missing

**Comment by david-allison-1** [comment]
at b82ec64db69ddf3bde8c211e4404ca6a4368e1bd

* git pr 9028 2
* PR + Cancel
* Make changes to UiUtil
* Rebuild

Same thing - all files missing

---


I tried again, and worked fine. Can't break it

**Comment by david-allison-1** [comment]
@ShridharGoel 

Is there a way that we can change this to allow the user to terminate the script at any time, without leaving a stash in a bad state?

It concerns me that we may remove a user's current commit with this change in its current state... but, it's much better than adding random CI errors to the middle of a PR

**Comment by ShridharGoel** [comment]
@david-allison-1 In the current script, stash is not being done.

**Comment by david-allison-1** [comment]
Oh, my issue was likely a residual from the previous script. Looks good to me then

**Review by david-allison-1 — APPROVED** [review]

**Review by mikehardy — APPROVED** [review]
Yes - after the file mode issue I was going to probe the git stash behavior as well, I was a little concerned that it was trying to maintain temporary state etc, and I'm glad there's a different strategy in play now. This looks great!

**Comment by mikehardy** [comment]
Clean commit too :-)

**Pull request merged**